src/core/DataFile.cpp

```diff
@@ -1386,6 +1387,85 @@ void DataFile::upgrade_noHiddenClipNames()
 }
 
 
+/** \brief Note range has been extended to match MIDI specification
+ *
+ * The non-standard note range previously affected all MIDI-based instruments
+ * except OpulenZ, and made them sound an octave lower than they should (#1857).
+ */
+void DataFile::upgrade_extendedNoteRange()
+{
+	if (!elementsByTagName("song").item(0).isNull())
+	{
+		// Dealing with a project file, go through all the tracks
+		QDomNodeList tracks = elementsByTagName("track");
+		for (int i = 0; !tracks.item(i).isNull(); i++)
+		{
+			// Ignore BB container tracks
+			if (tracks.item(i).toElement().attribute("type").toInt() == 1) { continue; }
+
+			QDomNodeList instruments = tracks.item(i).toElement().elementsByTagName("instrument");
+			if (instruments.isEmpty()) { continue; }
+			QDomElement instrument = instruments.item(0).toElement();
+			// Raise the base note of every instrument by 12 to compensate for the change
+			// of A4 key code from 57 to 69. This ensures that notes are labeled correctly.
+			instrument.parentNode().toElement().setAttribute(
+				"basenote",
+				instrument.parentNode().toElement().attribute("basenote").toInt() + 12);
+			// Raise the pitch of all notes in patterns assigned to instruments not affected
+			// by #1857 by an octave. This negates the base note change for normal instruments,
+			// but leaves the MIDI-based instruments sounding an octave lower, preserving their
+			// pitch in existing projects.
+			if (instrument.attribute("name") != "zynaddsubfx" &&
+				instrument.attribute("name") != "vestige" &&
+				instrument.attribute("name") != "lv2instrument" &&
+				instrument.attribute("name") != "carlapatchbay" &&
+				instrument.attribute("name") != "carlarack")
+			{
+				QDomNodeList patterns = tracks.item(i).toElement().elementsByTagName("pattern");
+				for (int i = 0; !patterns.item(i).isNull(); i++)
+				{
+					QDomNodeList notes = patterns.item(i).toElement().elementsByTagName("note");
+					for (int i = 0; !notes.item(i).isNull(); i++)
+					{
+						notes.item(i).toElement().setAttribute(
+							"key",
+							notes.item(i).toElement().attribute("key").toInt() + 12
+						);
+					}
+				}
+			}
+		}
+	}
+	else
+	{
+		// Dealing with a preset, not a song
+		QDomNodeList presets = elementsByTagName("instrumenttrack");
+		if (presets.item(0).isNull()) { return; }
+		QDomElement preset = presets.item(0).toElement();
+		// Common correction for all instrument presets (make base notes match the new MIDI range).
+		// NOTE: Many older presets do not have any basenote defined, assume they were "made for 57".
+		// (Specifying a default like this also happens to fix a FileBrowser bug where previews of presets
+		// with undefined basenote always play with the basenote inherited from previously played preview.)
+		bool ok = false;
+		int oldBase = preset.attribute("basenote").toInt(&ok);
+		if (!ok) { oldBase = 57; }
```

Comment:
```suggestion
		int oldBase = preset.attribute("basenote", "57").toInt();
```

I think that's slightly cleaner

Reply:
Thanks, I wasn't aware that you can pass a default value to the `attribute()`.